Added the absolute state of KDE software that should get released in lockstep with the KDE megarelease (#9)
Greybeard-Entertainment/pages: docs/are_we_gui_yet.org

#+TITLE: Are We GUI Yet
#+AUTHOR: Aleksandr Petrosyan

> This is a tiny discussion of why if you're writing your GUI in rust, life is

Reviewer: ```suggestion
This is a tiny discussion of why if you're writing your GUI in Rust, life is
```

difficult for you.

* Notes on Qt

  It has problems.

>   Have to fuck around with the build system.

Reviewer: ```suggestion
  For starters, you have to fuck around with the build system.
```

    cargo:rerun-if-env-changed=QMAKE
    cargo:rerun-if-env-changed=QT_VERSION_MAJOR
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=Qt6Core
    cargo:rustc-link-lib=Qt6Network
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=Qt6Core
    cargo:rustc-link-lib=Qt6Qml
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=Qt6Network
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=Qt6Core
    cargo:rustc-link-lib=Qt6Gui
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=Qt6Core
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=GLX
    cargo:rustc-link-search=/usr/lib
    cargo:rustc-link-lib=OpenGL
    cargo:rerun-if-changed=/home/app/Git/universal_config_gui/src/cxxqt_object.rs

    --- stderr
    thread 'main' panicked at /home/app/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cxx-qt-gen-0.5.3/src/syntax/qtfile.rs:44:57:
    Could not read path {} to string: Os { code: 2, kind: NotFound, message: "No such file or directory" }
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  warning: build failed, waiting for other jobs to finish...
  #+END_SRC

>   Where in this error message can you find mention of either the missing file:

Reviewer: Could you add a footnote (not necessarily in this PR) that gives some hint on determining a real filename, or is it just the use of search engines and forums? Or, maybe, a piece of the code/Makefile/whatever contains something pointing towards a proper name?

While I think I should know this already, I do remember many cases where it was quite different for different packages, and I had no automatic way to do that. I understand the idea that the change needs to be made, but I'm thinking in short-term here.

  =cxxqt_object.rs=, or the file that was actually present but wasn't supposed
  to be: =q_object.rs=.   Why bother with Rust at this point, if you're going
  to get shitty Qt library feedback might as well remove the extra layer of
  indirection

>   Have to fuck around with the Qt abstractions rather than circumvent them in a

Reviewer: ```suggestion
  Plus, you have to fuck around with the Qt abstractions rather than circumvent them in a
```

  Rust-idiomatic way.

>   Error handling.  Every function is both potentially fallible and never

Reviewer: ```suggestion
  Error handling is not any easier.  Every function is both potentially fallible and never
```

	    if let Some(engine) = engine.as_mut() {
		    engine.load(&QUrl::from("qrc:/main.qml"));
	    }

	    // Start the app
	    if let Some(app) = app.as_mut() {
		    app.exec();
	    }
    }
  #+END_SRC

  Engine returns an =Option= which is idiomatic Rust.  Not my preferred style,
  but I can forgive that.

  In Rust all lines that can fail, do so in one of two ways: either a panic,
  which can be controlled, but usually results in a traceback and termination,
  or by forwarding a =Result=.  Guess which of these does =engine.load= do?
  Well, it locks up the process, and prints a message that it couldn't find the
  file in the =qrc=.

  Now I've done this before, so I know what was missing.  But imagine that I
  didn't.  What, in your error message indicates what I need to do?

  Second problem is in the form of =app.exec=.  This is a blocking call.  More
  importantly this is a blocking call that /must/ live on the main thread.  Do
  you know what this means?  Exactly, it means that if you try and use =tokio=
  with this thing, you are going to find your life miserable.  See above.

  But to be quite honest, these things are problematic, but can be chalked up

>   to a skill issue.  WHat about the potential productivity?

Reviewer: ```suggestion
  to a skill issue.  What about the potential productivity?
```

  Well, again, I would argue that exposing your Rust code via =extern "C"= is

>   likely going to be more productive.  Inheritance, the C++ macros are

Reviewer: I think the sentence about the inheritance should be paraphrased to be more clear.

  implicit, shall we say...  baked into Qt.  It cannot exist without those
  features existing in other languages.

  And that's the key problem.

>   I get almost none of the benefits of Rust, because of the exposure to Qt.

Reviewer: ```suggestion
  I get almost none of the benefits of Rust, because of its exposure to Qt.
```

(I'm not sure how applicable it is.)

  There's no way of combining the two APIs.  There's no way to circumvent the
  meta-object-compiler, and Rust is compiled to something that has all of the

>   features of a C++ class.  Just use C++ then.  C++ isn't that terrible, and it

Reviewer: ```suggestion
  features of a C++ class.  At this point, using C++ becomes easier:  C++ isn't that terrible, and it
```

  Alas, because I had already tried to implement a parser in C++, so I will try
  and take the second approach.  I still think that contributing to the C++
  ecosystem via Qt applications using the native C++ is probably something I'll
  keep doing.  However, I will do so as part of pre-existing KDE projects, not
  by creating a new project.

* GTK on Relm4

  This is problematic.  I find that the amount of boilerplate needed to
  accommodate the stupidity of GTK also not doing things right in the exact
  same way as Qt is just too much for me.

  It's possible to do that.  But I sincerely doubt it'd be a useful use of my
  time.  Generating an ad-hoc component requires a "factory", (which is a sign
  of abstraction creep).

  Plain GTK4 is a problematic library.  For one simple reason.  The library
  breaks, and it requires dynamic linkage.  As such, this results in the
  program being potentially unmaintainable, if one does not wish to spend an
  inordinate amount of time fixing problems on platforms they don't even have.

  One way to fix it is to do static linkage.  The other way is to package into
  a reproducible format (/e.g./ =nix=, or =guix=).  Finally as a last resort,

>   one can package the problem into slightly (and only slightly) less defective

Reviewer: less defective compared to what?

  packaging standards, snaps and flatpaks.

* Iced
  This is raw.  But it is a forward-looking time investment.  At the very least
  it is actively developed and statically linked.  

  Let me explain why those two points are extremely important.  Firstly, and
  most importantly, there's no issue of outdated system libraries. Yes, you're
  wasting a bit more space that way, but far less than if you actually used a
  Flatpak and lost CPU cycles/had to do a bunch of =unsafe= to work around FFI.

  Secondly, the static linkage allows me to go into the implementations of the
  native components and write a new one in Rust. If that is not important,
  consider how much work would I need to do in order to do the same  with... I
  don't know, maybe Qt components. It is a foregone conclusion that any real
  implementation of a new Qt component is going to merely expose a QML
  interface, but be implemented in C++ and highly optimised.  Now consider that

>   it is not even not standard C++ it is not even the latest standard /of/ C++

Reviewer: Please elaborate on the "not even not" part.
